### PR TITLE
[Core] Edge Coupling Integration

### PR DESCRIPTION
**Pull request by tteschemacher**
**Description**
This PR aims to provide the possibility of integrating coupled geometries of 1 dimension. Generally, this can imply nurbs but also linear geometries or polygons.

To achieve this general implementation, the CurveTessellation was modified significantly.

**Changelog**
- Add CreateIntegrationPoints and Create QuadraturePointGeometries to CouplingGeometry.
- Modified CurveTessellation for more general purposes.

**Dependencies**
- #8490 SpansLocalSpace.
- #8491 ClosestPointLocalSpace.
- #8510 Integration Info.
- #8595 Refactor of integration

This deprecates #7016 .

### CODE REVIEW

**Comment by tteschemacher** [comment]
@dagmawibekel maybe you could check if this modification does affect your efficiency?

From the other side you will be able to use the tessellation with the parameter flag and can avoid the static cast which might be a really expensive operation.

**Comment by oberbichler** [comment]
Out of interest: what changes have been made to the tessellation algorithm aside from changing names?

**Comment by tteschemacher** [comment]
> Out of interest: what changes have been made to the tessellation algorithm aside from changing names?

I did not even change the names just the indentation ;)

- Basically, I made them generic to any type of geometry - which is an arguable advantage as all other existing curves do not need to be tessellated. However, like this, one can avoid the static_cast if the geometry is only known by Id. This is of advantage as we have different types of nurbs/brep curves.
- The second change is also arguable better: I changed the global coordinates type from Vector to array_1d<double, 3> - which is the coordinates type through all KRATOS geometries. This facilitates interfaces, however introduces not needed values. Efficiency wise this should however be the same as the dynamic vector has also more memory (e.g. size of vector).
- I added a function which splits the curve into similar segments, which is needed for e.g. initial guesses with axis intersections.
- I added a really bad function to access the closest point (iterating through all points). Not sure if you might have a more efficient algorithm for this? Would be more than welcomed.

File: kratos/geometries/coupling_geometry.h [hunk 1]
```diff
@@ -337,6 +304,230 @@ template<class TPointType> class CouplingGeometry
         return mpGeometries[0]->Center();
     }
 
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+    /// Provides the default integration per geometry.
+    virtual IntegrationInfo GetDefaultIntegrationInfo() const override
```
Comment (dagmawibekel) [inline kratos/geometries/coupling_geometry.h]: this shouldn't be virtual, right?

File: kratos/geometries/coupling_geometry.h [hunk 2]
```diff
@@ -337,6 +304,230 @@ template<class TPointType> class CouplingGeometry
...
+            std::vector<double> all_intersections;
+            MergeSpans(all_intersections, intersection_master_spans, intersection_master_spans_on_master_spans);
+
+            IntegrationPointUtilities::CreateIntegrationPoints1D(rIntegrationPoints, all_intersections, rIntegrationInfo.GetNumberOfIntegrationPointsPerSpan(0));
+        }
+    }
+
+    /* @brief This method creates a list of quadrature point geometries
+     *        from a list of integration points. It creates the list of
+     *        integration points byitself.
```
Comment (dagmawibekel) [inline kratos/geometries/coupling_geometry.h]: typo

File: kratos/geometries/coupling_geometry.h [hunk 3]
```diff
@@ -337,6 +304,230 @@ template<class TPointType> class CouplingGeometry
...
+            integration_points_global_coords_vector[i] = master_quadrature_points[i].Center();
+        }
+
+        // First slave
+        IntegrationPointsArrayType integration_points_slave = rIntegrationPoints;
+        CoordinatesArrayType local_slave_coords = ZeroVector(3);
+        CoordinatesArrayType global_slave_coords = ZeroVector(3);
+
+        bool use_tessellation = true;
+        if (use_tessellation) {
```
Comment (dagmawibekel) [inline kratos/geometries/coupling_geometry.h]: Can you explain this quickly?

File: kratos/geometries/coupling_geometry.h (hunk 1 above)
Comment (tteschemacher) [reply kratos/geometries/coupling_geometry.h]: thanks, is changed! 👍

File: kratos/geometries/coupling_geometry.h (hunk 2 above)
Comment (tteschemacher) [reply kratos/geometries/coupling_geometry.h]: the whole comment is a bit dodgy haha. Changed it.

Thanks!

File: kratos/geometries/coupling_geometry.h (hunk 3 above)
Comment (tteschemacher) [reply kratos/geometries/coupling_geometry.h]: That's good hard coding...

Basically, the idea is that you create a tessellation to find good initial guesses for the mapping towards the slaves.  Some geometries, however would ideally use different approaches. Not sure yet, how and if we will modify this. Additionally, triggering this isn't possible, yet. Probably, the integration info will be capable of containing the according data.

**Comment by tteschemacher** [comment]
> > I made them generic to any type of geometry
> > I changed the global coordinates type from Vector to array_1d<double, 3>.
> 
> Well, I am not aware of the Kratos-specific details and can therefore only give hints on the algorithm itself: I have implemented the original algorithm so that it can work with arbitrary curves. I do not know whether this was limited by the Kratos interface but it definitely makes sense to be generic. Recently I made some improvements which are only interesting for other types of geometries. For NURBS the current version works fine (polyline = special case of NURBS).

Well, I think in the way towards KRATOS the generality feature got lost a little bit. As I remember your code is better in that sense as you template the dimensional aspect, which makes you getting rid of the overhead coming from the vector. However, here we cannot really do that, so using the CoordinatesArrayType which is a array_1d is probably not the worst option.

Would be cool to check that algorithm. As far as I know, there is however, no one using different shaped objects.

> 
> > I added a function which splits the curve into similar segments, which is needed for e.g. initial guesses with axis intersections.
> 
> Which function?

`static TessellationType ComputePolygon`
I just realized this was already merged. It is line 233ff. It basically makes use of the same data structure but a different algorithm, simpler and also for different use cases.

> 
> > I added a really bad function to access the closest point (iterating through all points).
> 
> Basically that's the way to go. I improved the spatial search in terms of speed but without changing the quality of the results. Can be done in a second step.

The problem here is that yet I create this quite some times, and creating a spatial tree every time would be a bit a killer. But I totally agree that this is something important for the future. Would be good to chat about your approach..

Thanks a lot for your comments @oberbichler . Really highly appreciated!

**Comment by oberbichler** [comment]
> `static TessellationType ComputePolygon`

OK but keep in mind that such an algorithm might be unstable for real-world problems.

> creating a spatial tree every time would be a bit a killer.

That is the reason why I have implemented these algorithms as classes and not as functions.

No blocker. Both things can be improved/fixed with future PRs.

File: kratos/geometries/coupling_geometry.h [hunk 4]
```diff
@@ -337,6 +304,236 @@ template<class TPointType> class CouplingGeometry
...
+
+            mpGeometries[0]->SpansLocalSpace(intersection_master_spans);
+            CurveTessellation<PointerVector<TPointType>> curve_tessellation;
+            curve_tessellation.Tessellate(
+                *(mpGeometries[0].get()),
+                intersection_master_spans,
+                1e-2, mpGeometries[0]->PolynomialDegree(0), false);
+
+            CoordinatesArrayType local_coords_slave = ZeroVector(3);
+            CoordinatesArrayType global_coords = ZeroVector(3);
```
Comment (manuelmessmer) [inline kratos/geometries/coupling_geometry.h]: Should be global_coords_slave, I guess.

File: kratos/geometries/coupling_geometry.h [hunk 5]
```diff
@@ -337,6 +304,236 @@ template<class TPointType> class CouplingGeometry
...
+
+                for (IndexType j = 0; j < intersection_slave_spans.size(); ++j) {
+                    local_coords_slave[0] = intersection_slave_spans[j];
+                    mpGeometries[i]->GlobalCoordinates(
+                        global_coords, local_coords_slave);
+                    curve_tessellation.GetClosestPoint(
+                        global_coords, global_coords_master, local_coords_master);
+                    int success = mpGeometries[0]->ProjectionPoint(
+                        global_coords, global_coords_master, local_coords_master);
+                    KRATOS_DEBUG_ERROR_IF(success == 1 && (norm_2(global_coords - global_coords_master) > model_tolerance))
```
Comment (manuelmessmer) [inline kratos/geometries/coupling_geometry.h]: Is "&&" correct or "||"?

File: kratos/geometries/coupling_geometry.h [hunk 6]
```diff
@@ -337,6 +304,236 @@ template<class TPointType> class CouplingGeometry
...
+        mpGeometries[0]->CreateQuadraturePointGeometries(
+            master_quadrature_points,
+            NumberOfShapeFunctionDerivatives,
+            rIntegrationPoints,
+            rIntegrationInfo);
+
+        // Compute vector of location
+        CoordinatesArrayVectorType integration_points_global_coords_vector(num_integration_points);
+        for (SizeType i = 0; i < num_integration_points; ++i) {
+            integration_points_global_coords_vector[i] = master_quadrature_points[i].Center();
```
Comment (manuelmessmer) [inline kratos/geometries/coupling_geometry.h:431]: Do you mean the center of the geometry or the coordinates of the quadrature point?

File: kratos/geometries/coupling_geometry.h [hunk 7]
```diff
@@ -337,6 +304,236 @@ template<class TPointType> class CouplingGeometry
...
+                    int success = mpGeometries[0]->ProjectionPoint(
+                        global_coords, global_coords_master, local_coords_master);
+                    KRATOS_DEBUG_ERROR_IF(success == 1 && (norm_2(global_coords - global_coords_master) > model_tolerance))
+                        << "Projection of intersection spans failed. Global Coordinates on slave: "
+                        << global_coords << ", and global coordinates on master: "
+                        << global_coords_master << ". Difference: " << norm_2(global_coords - global_coords_master)
+                        << " larger than model tolerance: " << model_tolerance << std::endl;
+
+                    // If success == 0, it is considered that the projection is on one of the boundaries.
+                    intersection_master_spans_on_master_spans.push_back(local_coords_master[0]);
```
Comment (manuelmessmer) [inline kratos/geometries/coupling_geometry.h]: master_spans_on_master_spans???

File: kratos/geometries/coupling_geometry.h [hunk 8]
```diff
@@ -337,6 +304,236 @@ template<class TPointType> class CouplingGeometry
...
+
+        // First slave
+        IntegrationPointsArrayType integration_points_slave = rIntegrationPoints;
+        CoordinatesArrayType local_slave_coords = ZeroVector(3);
+        CoordinatesArrayType global_slave_coords = ZeroVector(3);
+
+        if (!rIntegrationInfo.Is(IntegrationInfo::DO_NOT_CREATE_TESSELLATION_ON_SLAVE)) {
+            if (this->Dimension() == 1) {
+                CurveTessellation<PointerVector<TPointType>> curve_tesselation;
+                curve_tesselation.Tessellate(*(mpGeometries[1].get()), 1e-2, mpGeometries[1]->PolynomialDegree(0));
```
Comment (manuelmessmer) [inline kratos/geometries/coupling_geometry.h:442]: Do you always only have one slave, becaus in CreateQuadrature you have a loop.

File: kratos/geometries/coupling_geometry.h [hunk 9]
```diff
@@ -337,6 +304,236 @@ template<class TPointType> class CouplingGeometry
...
+    }
+
+    static void SortUnique(
+        std::vector<double>& rResultSpans,
+        const double Tolerance)
+    {
+        std::sort(std::begin(rResultSpans), std::end(rResultSpans));
+
+        auto last = std::unique(std::begin(rResultSpans), std::end(rResultSpans),
+            [=](double a, double b) { return b - a < Tolerance; });
```
Comment (manuelmessmer) [inline kratos/geometries/coupling_geometry.h]: Aren't you capturing here the vector as well (by value)? Not only the tolerance?

Comment (tteschemacher) [reply kratos/geometries/coupling_geometry.h]: Not sure what you mean. It works like this ;)

File: kratos/geometries/coupling_geometry.h (hunk 8 above)
Comment (tteschemacher) [reply kratos/geometries/coupling_geometry.h:442]: Yes, that is true.

Create integration points is capable to capture multiple curves. Whereas, up to date, CreateQuadraturePoints is not. This, however can be extended, once needed.

File: kratos/geometries/coupling_geometry.h (hunk 7 above)
Comment (tteschemacher) [reply kratos/geometries/coupling_geometry.h]: Yep. Renamed it.

File: kratos/geometries/coupling_geometry.h (hunk 6 above)
Comment (tteschemacher) [reply kratos/geometries/coupling_geometry.h:431]: The center of a quadrature point does provide the global location of it. Yet, it is the most efficient way to access its location.

File: kratos/geometries/coupling_geometry.h (hunk 5 above)
Comment (tteschemacher) [reply kratos/geometries/coupling_geometry.h]: We will need to gain more experience about this check in the future. 

For the current implementation it makes sense to allow failed points which are within the model tolerance

File: kratos/geometries/coupling_geometry.h (hunk 9 above)
Comment (manuelmessmer) [reply kratos/geometries/coupling_geometry.h]: I mean, it should be enough if you only capture the Tolerance variable in [].

File: kratos/geometries/coupling_geometry.h (hunk 4 above)
Comment (tteschemacher) [reply kratos/geometries/coupling_geometry.h]: I renamed some variables. Hope it is more consistent now?

File: kratos/utilities/tessellation_utilities/curve_tessellation.h [hunk 10]
```diff
@@ -254,23 +307,86 @@ class NurbsCurveTessellation
...
+    ///@{
+
+    static double DistanceToLine(
+        const typename GeometryType::CoordinatesArrayType& rPoint,
+        const typename GeometryType::CoordinatesArrayType& rLineA,
+        const typename GeometryType::CoordinatesArrayType& rLineB
+    )
+    {
+        typename GeometryType::CoordinatesArrayType vector_v = rLineA - rPoint;
+        typename GeometryType::CoordinatesArrayType vector_u = rLineB - rLineA;
```
Comment (manuelmessmer) [inline kratos/utilities/tessellation_utilities/curve_tessellation.h]: might be worth to define at the top.

File: kratos/utilities/tessellation_utilities/curve_tessellation.h [hunk 11]
```diff
@@ -254,23 +307,86 @@ class NurbsCurveTessellation
...
+
+    /* @brief This method iterates through all points within the
+     *        polygon and returns the closest point.
+     *
+     * @param CoordinatesArrayType external point
+     * @param rClosestPointGlobalCoordinates closest point within polygon in global coordinates.
+     * @param rClosestPointLocalCoordinates closest point within polygon in local coordinates.
+     * @param rTesselation corresponding tessellation.
+     */
+    void GetClosestPoint(
```
Comment (manuelmessmer) [inline kratos/utilities/tessellation_utilities/curve_tessellation.h:346]: why do we need this a static and non-static?

File: kratos/utilities/tessellation_utilities/curve_tessellation.h [hunk 12]
```diff
@@ -254,23 +307,86 @@ class NurbsCurveTessellation
...
+        const CoordinatesArrayType& rPointGlobalCoordinates,
+        CoordinatesArrayType& rClosestPointGlobalCoordinates,
+        CoordinatesArrayType& rClosestPointLocalCoordinates) const
+    {
+        GetClosestPoint(
```
Comment (manuelmessmer) [inline kratos/utilities/tessellation_utilities/curve_tessellation.h:351]: If this function is const, shuldn't the one that is called here also be const?

**Review by manuelmessmer — COMMENTED** [review]
@tteschemacher  some minor comments, otherwise look good to me now. Then let"s do curve in volume! :rocket:

File: kratos/utilities/tessellation_utilities/curve_tessellation.h (hunk 10 above)
Comment (tteschemacher) [reply kratos/utilities/tessellation_utilities/curve_tessellation.h]: 👍

File: kratos/utilities/tessellation_utilities/curve_tessellation.h (hunk 12 above)
Comment (tteschemacher) [reply kratos/utilities/tessellation_utilities/curve_tessellation.h:351]: the other function is static, which is implying const. Thusly, it is running with this.

File: kratos/utilities/tessellation_utilities/curve_tessellation.h (hunk 11 above)
Comment (tteschemacher) [reply kratos/utilities/tessellation_utilities/curve_tessellation.h:346]: Creating a tessellation is a bit a costly function, which accordingly might be worth to safe from time to time. Accordingly is the duplicated interface.

**Review by manuelmessmer — APPROVED** [review]
Beautiful. Looks good to me now.

**Comment by tteschemacher** [comment]
:octocat:  ➿ 🥳

**Pull request merged**